.circleci/config.yml

```diff
@@ -102,45 +57,32 @@ jobs:
           command: |
             brew install pyenv
 
-#      - restore_cache:
-#          keys:
-#            - pyenv-{{ .Environment.CIRCLE_JOB }}
-
       - run:
           name: install python
           command: |
             pyenv install $PYTHON -s
 
-#      - save_cache:
-#          paths:
-#            - ~/.pyenv
-#          key: pyenv-{{ .Environment.CIRCLE_JOB }}
-
-#      - restore_cache: *restore-cache-template
-
+      #TODO: add caching back in when demos have the same requirements and exist in the same venv
       - run:
           name: create virtualenv
           command: |
             eval "$(pyenv init -)"
             pyenv local $PYTHON
             python -m pip install virtualenv
-            python -m virtualenv env
 
       - run:
-          name: install dependencies
-          command: |
-            . env/bin/activate
-            pip install -r requirements_tests.txt
-
-#      - save_cache: *save-cache-template
-
-      - run: &run-tests-template
-          name: run tests
+          name: install dependencies and run tests
           command: |
-            . env/bin/activate
-            coverage run -m unittest discover
-            deactivate
-#      - run: *run-tests-template
+            demos=$(ls -d */ | grep -v "tests" | grep -v "circuit-fault-diagnosis" | sed 's/\///')
```

Comment:
See [bash pitfall #1](https://mywiki.wooledge.org/BashPitfalls#for_f_in_.24.28ls_.2A.mp3.29).

You're naively assuming directory names are Latin letters without whitespace characters (like spaces, tabs, newlines). Or that `*` can be a directory name. Also your regexes are not anchored.

Check out `find`, and in particular `-print0` action.

Also, assuming our CI runs Bash 4 (true by default everywhere except on Macs), you can use extended globbing instead (see `shopt` and `extglob` option) to replace that complete pipeline.

Reply:
Thank you for the link and for the detailed reasoning on why I shouldn't do this. Will make the fix now.

Reply:
You're very welcome.